common/math/math.go

```diff
@@ -34,3 +36,165 @@ func RoundFloat(x float64, prec int) float64 {
+func CalculateCompoundAnnualGrowthRate(openValue, closeValue, intervalsPerYear, numberOfIntervals float64) float64 {
+	k := math.Pow(closeValue/openValue, intervalsPerYear/numberOfIntervals) - 1
+	return k * 100
+}
+
+// CalculateCalmarRatio is a function of the average compounded annual rate of return versus its maximum drawdown.
+// The higher the Calmar ratio, the better it performed on a risk-adjusted basis during the given time frame, which is mostly commonly set at 36 months
+func CalculateCalmarRatio(highestPrice, lowestPrice, average float64) float64 {
+	if highestPrice == 0 {
+		return 0
+	}
+	drawdownDiff := (highestPrice - lowestPrice) / highestPrice
+	if drawdownDiff == 0 {
+		return 0
+	}
+	return average / drawdownDiff
+}
+
+// CalculateInformationRatio The information ratio (IR) is a measurement of portfolio returns beyond the returns of a benchmark,
+// usually an index, compared to the volatility of those returns.
+// The benchmark used is typically an index that represents the market or a particular sector or industry.
+func CalculateInformationRatio(values, benchmarkRates []float64, averageValues, averageComparison float64) float64 {
+	if len(benchmarkRates) == 1 {
+		for i := range values {
+			if i == 0 {
+				continue
+			}
+			benchmarkRates = append(benchmarkRates, benchmarkRates[0])
+		}
+	}
+	var diffs []float64
+	for i := range values {
+		diffs = append(diffs, values[i]-benchmarkRates[i])
+	}
+	stdDev := PopulationStandardDeviation(diffs)
+	if stdDev == 0 {
+		return 0
+	}
+	return (averageValues - averageComparison) / stdDev
+}
+
+// PopulationStandardDeviation calculates standard deviation using population based calculation
+func PopulationStandardDeviation(values []float64) float64 {
+	if len(values) == 0 {
+		return 0
+	}
+	avg := ArithmeticAverage(values)
+	diffs := make([]float64, len(values))
+	for x := range values {
+		diffs[x] = math.Pow(values[x]-avg, 2)
+	}
+	return math.Sqrt(ArithmeticAverage(diffs))
+}
+
+// SampleStandardDeviation standard deviation is a statistic that
+// measures the dispersion of a dataset relative to its mean and
+// is calculated as the square root of the variance
+func SampleStandardDeviation(vals []float64) float64 {
+	if len(vals) <= 1 {
+		return 0
+	}
+	mean := ArithmeticAverage(vals)
+	var superMean []float64
+	var combined float64
+	for i := range vals {
+		result := math.Pow(vals[i]-mean, 2)
+		superMean = append(superMean, result)
+		combined += result
+	}
+	avg := combined / (float64(len(superMean)) - 1)
+	return math.Sqrt(avg)
+}
+
+// GeometricAverage is an average which indicates the central tendency or
+// typical value of a set of numbers by using the product of their values
+// The geometric average can only process positive numbers
+func GeometricAverage(values []float64) float64 {
+	if len(values) == 0 {
+		return 0
+	}
+	product := 1.0
+	for i := range values {
+		if values[i] <= 0 {
+			// cannot use negative or zero values in geometric calculation
```

Comment:
Unfortunately we can have negative values :D. It just means that all values need to be negative or all values need to be positive. Also probably should return err here too now that these paths are added and there will be no real way to distinguish what went wrong when just returning 0. Really picky on these but 🤷

Reply:
On second thought since it doesn't really mean too much to get geometric mean of negative numbers we dont need to add in support but rather just put out an error I guess. https://math.stackexchange.com/questions/1371236/geometric-mean-of-negative-numbers-is-positive-or-negative